### ISSUE TITLE
Unable to use open file menu to open zarr files

### ISSUE THREAD

**Issue by alisterburt** [napari/napari#5410]
## 🐛 Bug

Whilst implementing an IO plugin for files backed by zarr over at napari-threedee/napari-threedee#85 I noticed that I'm unable to open zarr files from the standard napari 'open file' menu because zarr files are themselves directories. I know we have the separate 'open folder' menu but the experience is pretty bad if a user wants to open their zarr file and can't...

Is there a good reason we keep opening files and folders separate in the GUI? 🙂 

cc @DragaDoncila and @Czaki for the plugin infra/qt sides of this

**Comment by Czaki** [comment]
> Is there a good reason we keep opening files and folders separately in the GUI? slightly_smiling_face

This is more intuitive to the user when clicking on the folder and does not try to open it but go into which is a common case for most of the file formats. 

I think that I may know how to handle this but only using Qt dialogs, not a native one. Maybe it could be done also using native dialogs, but then I do not know how to test it, and may require manual curation.

**Comment by alisterburt** [comment]
I agree clicking a folder should not open it (this doesn't happen even in the folder dialog).

What I'm suggesting is that clicking the open button in the GUI when a folder is selected in open file mode should attempt to open that folder in napari rather than going into the folder in the dialog

**Comment by Czaki** [comment]
And most users will see error information about lack of reader. 

But maybe we could add an appropriate field in settings?

In general, I know how to do it properly using qt dialog (not a native one). Then we could check reader presence based on a change of selection. So we could dynamically control behavior and have the possibility to provide user a proper information.

So maybe we could have settings option to check if we want native dialog (better system integration) or at based with better feedback from interface.

**Comment by alisterburt** [comment]
I don't think that error would come up often enough to be a problem for people, maybe I'm wrong? I only see people hitting the open button when they want to open a file, not to navigate into a folder in the dialog - open to settings but don't think it would be used much in practice

That would be awesome, I think we can achieve this easily with the native dialogs too though, given that the behaviour of the native 'open folder' dialog we already have is perfect, it's just that it makes files impossible to select

**Comment by DragaDoncila** [comment]
@alisterburt this has been oft discussed and always the push back has been that it could not be done with a native dialog. It can definitely be done with a `QtFileDialog`. I think it's worth revisiting. There's lots to improve about this.

See #4612 where this was last proposed. My feeling is that we should move to one `File - Open` dialog where the user can browse for files, choose the plugin to open with, select whether they want to stack or not, etc.

**Comment by dstansby** [comment]
On macOS and napari 0.4.19 I can open a zarr folder fine using the "Open folder" menu item, so this can probably be closed?

**Comment by psobolewskiPhD** [comment]
I think that was also possible when the issue was opened. I think it's more about the confusion -- I don't think a lot of people may know a .zarr file is a folder. It's not obvious at all. 
So the title of the issue is wrong maybe and should focus on this part:
> Is there a good reason we keep opening files and folders separate in the GUI? 🙂

And to make it worse, sometimes you do need to navigate into the .zarr to open it (e.g. ome-zarr)
And then there is the issue of the way that native file handling is...

**Comment by Czaki** [comment]
> So the title of the issue is wrong maybe and should focus on this part:
> 
> > Is there a good reason we keep opening files and folders separate in the GUI? 🙂

The technical reason is using native dialog

**Comment by psobolewskiPhD** [comment]
Right, but even using a non-native dialog solving the issue of sometimes wanting to open:
foo.zarr -- like a file
and sometimes
foo.zarr/0/ -- like a folder

would not be trivial, UX-wise. Especially given that you can't glean what you need a priori, they both are .zarr

**Comment by Czaki** [comment]
If we ere not using native file dialog, we could change behavior based on file type selector.

And `foo.zarr/0/` do not end with `.zarr` so zarr reader should not be triggered.

**Comment by DragaDoncila** [comment]
Yeah I wouldn't close this yet until we solve the bigger issue of the very complicated file opening workflow, as detailed by Jun in #4612, discussed by me in the context of writing in #6653, and many other places. It's on my todo list to write a roadmap item for a richer and smoother import/export experience. Clearly I need to get on that...

**Comment by Czaki** [comment]
I prefer to have one Issue that will trace all others to simplify cleaning when it is finally resolved.

**Comment by DragaDoncila** [comment]
@Czaki ok then perhaps when I go through and make the roadmap issue I'll close any relevant issues I refer to (like this one) and direct focus back to that one? @psobolewskiPhD what do you think?

**Comment by psobolewskiPhD** [comment]
Sounds good to me.

**Comment by psobolewskiPhD** [comment]
I will note that on macOS .zarr are indicted in the file browser (Finder) as directories -- because they are directories. 
I recently helped a user on Linux get started with napari to access/proof-read zarrs and introduced them as directories and they immediately used the File > Open Folder menu item.

Also, i just noticed:

> And foo.zarr/0/ do not end with .zarr so zarr reader should not be triggered.

In fact it should, because /0 is likely to be a valid zarr either an array (so an image) or a group of groups or arrays.
the .zarr is just a container for things. One can very frequently want to open a specific array (image) like:
`foo.zarr/0/2`
foo.zarr is the root of the hierarchy, it's a Group
foo.zarr/0 is a Group within foo.zarr, this could be a collection of images, e.g. a multiscale image
foo.zarr/0/2 is a specific resolution level
(note, all of those all directories)
OME-zarr has metadata in a /OME directory too, but "vanilla" zarr can have arbitrary hierarchies inside, not limited to /0 or /0/0. It will be very hard to write a reader that can handle anything, so users will need to be able to navigate and open things within a .zarr, but they will always be opening directories to load the array data in the end.

Some of this already works, but i will be making a PR to try to improve handling a bit.
